src/scheduling/cost_models/simulated_quincy_cost_model.cc

```diff
@@ -0,0 +1,358 @@
+
+// The cost from the unscheduled to the sink is 0. Setting it to a value greater
+// than zero affects all the unscheduled tasks. It is better to affect the cost
+// of not running a task through the cost from the task to the unscheduled
+// aggregator.
+Cost_t SimulatedQuincyCostModel::UnscheduledAggToSinkCost(JobID_t job_id) {
+  return 0ULL;
+}
+
+// The cost from the task to the cluster aggregator models how expensive is a
+// task to run on any node in the cluster. The cost of the topology's arcs are
+// the same for all the tasks.
+Cost_t SimulatedQuincyCostModel::TaskToClusterAggCost(TaskID_t task_id) {
+	return cluster_aggregator_cost_[task_id];
+}
+
+Cost_t SimulatedQuincyCostModel::TaskToResourceNodeCost(TaskID_t task_id,
+                                               ResourceID_t resource_id) {
+  ResourceCostMap_t *cost_map = preferred_machine_map_[task_id];
+  return (*cost_map)[resource_id];
+}
+
+// Cost from machines to cores, always 0.
+Cost_t SimulatedQuincyCostModel::ResourceNodeToResourceNodeCost(
+    ResourceID_t source,
+    ResourceID_t destination) {
+  return 0ULL;
+}
+
+// The cost from the resource leaf to the sink is 0.
+Cost_t SimulatedQuincyCostModel::LeafResourceNodeToSinkCost(ResourceID_t resource_id) {
+  return 0ULL;
+}
+
+Cost_t SimulatedQuincyCostModel::TaskContinuationCost(TaskID_t task_id) {
+	// TODO(adam): task preemption support
+  return 0ULL;
+}
+
+Cost_t SimulatedQuincyCostModel::TaskPreemptionCost(TaskID_t task_id) {
+	// TODO(adam): task preemption support
+  return 0ULL;
+}
+
+Cost_t SimulatedQuincyCostModel::TaskToEquivClassAggregator(TaskID_t task_id,
+                                                   EquivClass_t tec) {
+  // task to rack aggregators
+  return preferred_rack_map_[task_id][tec];
+}
+
+Cost_t SimulatedQuincyCostModel::EquivClassToResourceNode(EquivClass_t tec,
+                                                 ResourceID_t res_id) {
+	// cost of arcs from rack aggregators are always zero
+	// (costs are instead encoded in arc from task to aggregator)
+	return 0LL;
+}
+
+Cost_t SimulatedQuincyCostModel::EquivClassToEquivClass(EquivClass_t tec1,
+                                                        EquivClass_t tec2) {
+	// this shouldn't be called; the only arcs are from cluster aggregator to rack
+	// aggregator, but Firmament considers cluster aggregator to be a special case
+	CHECK(false);
+  return 0LL;
+}
+
+// In Quincy, a task is in its own equivalence class
+vector<EquivClass_t>* SimulatedQuincyCostModel::GetTaskEquivClasses(
+    TaskID_t task_id) {
+  vector<EquivClass_t>* preferred_res = new vector<EquivClass_t>();
+  auto &preferred_racks = preferred_rack_map_[task_id];
+  for (auto elt : preferred_racks) {
+    EquivClass_t rack = elt.first;
+    VLOG(1) << "Task " << task_id << " has arc to rack aggregator " << rack;
+    preferred_res->push_back(rack);
+  }
+  return preferred_res;
+}
+
+vector<EquivClass_t>* SimulatedQuincyCostModel::GetResourceEquivClasses(
```

Comment:
Hmm..I think from this method we should just return the equiv_classes to which the resource node is going to be connected. Similarly, from the GetOutgoingEquivClassPrefArcs we should return the ResIDs to which the equiv_class will be connected (i.e., just the forward going arcs).